### PR TITLE
fix(node): validate git branch refs to close option injection

### LINKED ISSUE #378: Unvalidated PR/repo branch refs reach git as options (arbitrary file write via PR diff)
## Summary

`create_pr` stores the PR's `source_branch` and `target_branch` from the request body without validating them as git refs. Both are later interpolated into a single `git` argv element, so a branch name beginning with `-` is parsed by git as an option rather than a revision. A `target_branch` of `--output=<path>` turns the PR diff endpoint into an arbitrary file write with the node process's privileges.

## Impact

`GET /api/v1/repos/{owner}/{repo}/pulls/{n}/diff` builds `git diff {target}...{source}` as one argument (`branch_diff_names` / `branch_diff` in `crates/gitlawb-node/src/git/store.rs`). With `target_branch = --output=/tmp/x`, git runs `git diff --output=/tmp/x...{source}` and writes to that path. The same shape reaches `git worktree add _merge_worktree {target}` and `git merge --no-ff {source}` in `merge_branch`.

Reachability: opening the PR needs read access (`create_pr` gates on `authorize_repo_read`), but the diff endpoint takes an optional identity, so on a public repo the trigger is unauthenticated. The withhold check that builds the diff runs before the per-path visibility gate, so the write happens even when the request ends in a denial or a 500.

Confirmed by reproduction against the real handlers: an anonymous `GET .../pulls/1/diff` on a public repo, after a reader opened a PR with an `--output=` target, created the attacker-named file on disk.

## Fix

Validate branch refs at the `create_pr` storage boundary (the single point both sink families read from) against `git check-ref-format` rules, rejecting a leading `-` as the security-critical case. A `--` delimiter is not the fix: `{target}...{source}` is a revision argument, and `--` there reinterprets the revision as a pathspec.

### PR DESCRIPTION
**Pull request by beardthelion**
Closes #378.

PR branch refs and a repo's `default_branch` were stored from request bodies with no ref validation, then interpolated into single git argv elements downstream: `git diff {target}...{source}` in `branch_diff`/`branch_diff_names`, and `git worktree add ... {target}` / `git merge {source}` in `merge_branch`. A value starting with `-` is read by git as an option, so a stored `--output=<path>` turns the PR diff endpoint into an arbitrary file write. `get_pr_diff` takes an optional identity, so the trigger is unauthenticated on a public repo; planting the PR needs only read access, and the write happens at the withhold check before the visibility gate.

Two caller-supplied entry points feed those sinks:
- `create_pr` stores `source_branch` and `target_branch` from the body.
- `create_repo` stores `default_branch` from the body, which becomes a PR's `target_branch` when the PR omits one.

This adds a shared `validate_git_ref` (canonical in `git/store.rs`, re-exported as `crate::api::validate_git_ref`), following git check-ref-format rules with a leading-dash rejection as the core, and calls it at both boundaries. `create_pr` validates `source_branch` and the *resolved* `target_branch`, so a poisoned or legacy default cannot reach the sink even if it bypassed `create_repo`'s gate. `create_repo` validates `default_branch`. Both run after the existing auth/name checks and before the row is written. No `--` delimiter is added at the git call sites: the argument is a revision, and `--` there would reinterpret it as a pathspec.

Not affected: `resolve_head` prefixes `refs/heads/` before passing the branch to git, so a leading dash cannot lead there; `fork_repo` takes no branch from the request.

The guard runs at two layers. The storage boundaries above fail fast with a 400 and keep option-shaped refs out of the DB. The git sink functions (`branch_diff`, `branch_diff_names`, `merge_branch`) also reject them before building the argv, so the property holds for every caller and every row, including a row written before this change or by any writer that skips the boundary check.

## Tests
- `validate_git_ref` unit tests: accepts `main`, `feature/foo`, `release-1.2`, `v1.0.0`, `user/fix-bug`; rejects empty, `--output=/tmp/x`, `-rf`, a space, `..`, `~`, `@{`, `.lock`, leading/trailing/`//` slashes, and an over-255-byte name.
- End-to-end: `create_pr` as the owner of a public repo with an `--output=` target returns 400 with no PR row (fails before the fix, which returns 201); `create_repo` with an `--output=` `default_branch` returns 400 with no repo row.
- A poisoned PR row inserted directly (bypassing the handler) does not write a file when `get_pr_diff` runs, proving the sink guard, not just the boundary; it is RED without the sink guard.
- `create_pr` also rejects an option-shaped `source_branch` (the `git merge` arm).
- Full `gitlawb-node` suite: 829 passed, 0 failed.

## Summary by CodeRabbit

- **Bug Fixes**
  - Branch and Git reference names are now validated when creating repositories and pull requests.
  - Invalid, malformed, shorthand, pseudoref, or option-like references are rejected with clear bad-request responses.
  - Git execution failures now return appropriate Git errors.
  - Unsafe references are blocked during branch comparisons and merges, including previously stored values.
  - Comparisons and merges target the intended branches, report failures accurately, and require successful commit results.
  - Invalid requests no longer create records or report unsuccessful merges as successful.

- **Tests**
  - Added comprehensive security and merge regression coverage.

### CODE REVIEW

**Comment by kevincodex1** [comment]
@coderabbitai please review

**Review by jatmn — CHANGES_REQUESTED** [review]
I found issues that need to be addressed before this is ready.

## Merge readiness

- [P1] Rebase onto current `main` and re-review the resolved security diff
  `crates/gitlawb-node/src/git/store.rs:701`
  The PR merge base (`e4c7458`) is behind the live target (`73fd747`), which has changed the same node API and Git-store security surfaces. Rebase before merge so the option-injection fix is verified against the current authorization and subprocess behavior.

## Findings

- [P2] Validate branch names rather than accepting revision expressions
  `crates/gitlawb-node/src/git/store.rs:701`
  `validate_git_ref` is introduced as the shared gate for persisted PR `source_branch`/`target_branch` values and repository `default_branch` values, but it implements only part of Git’s branch-name grammar. In particular, it accepts `HEAD` and components ending in `.`, while `git check-ref-format --branch` rejects `HEAD`, `feature.`, and `feature/x.`.

  This is not just a validation-message mismatch. `create_pr` accepts these values, writes the row, and emits `pull_request.opened`. A trailing-dot name then makes the later `branch_diff_names`, `branch_diff`, or `merge_branch` invocation fail. More importantly, `HEAD` is a revision expression, not a branch: `merge_branch` checks out the target branch and then runs `git merge --no-ff HEAD`, which Git treats as the already checked-out target commit and exits successfully without merging a submitted source branch. The handler then resolves the target ref and records the PR as merged, emitting `pull_request.merged` despite no source-branch merge having occurred.

  Please address the root cause by making this one shared validator enforce the actual *branch-name* contract—not merely a subset sufficient to block leading options—and retain its use at both request storage boundaries and all legacy-row subprocess sinks. A Git-equivalent branch-format check (or a complete equivalent implementation) should reject symbolic revision names and every forbidden component form, while preserving ordinary branch names and the existing leading-dash protection. Add load-bearing boundary and sink tests for at least `HEAD`, `feature.`, and `feature/x.`: rejected API requests must create no row or webhook, and direct legacy rows must be rejected before any Git command or merge-side effect.

**Comment by beardthelion** [comment]
Rebased onto current main.

`validate_git_ref` now calls `git check-ref-format --branch` and also rejects `refs/` prefixes, so symbolic names (`HEAD`), trailing-dot components (`feature.`, `feature/x.`), option-shaped refs, and fully qualified refs like `refs/tags/v1` cannot be stored or reach the git sinks.

Load-bearing tests at both boundaries and sinks:
- `create_pr` / `create_repo` reject HEAD, trailing dots, `refs/tags/v1`, `refs/heads/main` (400, no row)
- Poisoned-row sink tests for diff (`--output=` target), merge (`--output=` source), and HEAD source/target at merge
- `create_pr` with omitted `target_branch` and a poisoned `default_branch` in the DB row (resolved target validation)

Head `799b73a`. Targeted suite green locally.

**Review by jatmn — CHANGES_REQUESTED** [review]
I found issues that need to be addressed before this is ready.

## Findings

- [P1] Reject Git revision shorthands rather than treating them as branch names
  `crates/gitlawb-node/src/git/store.rs:765`
  `git check-ref-format --branch @` succeeds, but that command only validates ref-name grammar; it does not prove that the string will be interpreted as a literal local branch at a later Git call. The validator retains the original `@` value, and every sink passes it as a bare revision argument. Git interprets it as `HEAD`. Consequently `create_pr` accepts and persists `source_branch: "@"`; after `merge_branch` checks out the target worktree, `git merge --no-ff @` is a successful no-op merge of that checked-out target. The handler then marks the PR merged and emits `pull_request.merged` even though no submitted branch was merged. `@{-1}` similarly selects a reflog/checkout-relative branch, and special pseudorefs have the same class of ambiguity.

  Please address the root cause: make the stored branch-field contract and each sink's Git arguments unambiguously refer to a local branch, rather than relying on a grammar check followed by a bare revision expression. That can mean rejecting every revision shorthand/pseudoref that cannot safely denote a branch, or constructing and using an explicit local-branch ref at the sinks after validating the name. Do not merely compare `check-ref-format` output to the input—`@` is accepted unchanged. Cover `@`, `@{-1}`, and relevant pseudorefs at both request boundaries and legacy-row diff/merge sinks, including assertions that a rejected merge leaves the PR open and emits no merge webhook.

- [P3] Preserve Git launch failures as server errors
  `crates/gitlawb-node/src/git/store.rs:773`
  The root cause is that `validate_git_ref` collapses two distinct outcomes into `Result<(), String>`: a completed `check-ref-format` process rejecting caller input, and an operational failure to spawn or collect output from Git. Both new storage boundaries then map that string to `AppError::BadRequest`. If Git is missing or cannot execute, valid repository and pull-request creation requests therefore receive HTTP 400 and are told to change their branch name instead of receiving the existing server-side Git failure response.

  Preserve that distinction in the validator's error type (or at each caller): only a completed validation rejection should become 400; spawn and command-output failures should flow to the existing Git/server-error path. Add a focused test that exercises the operational-error branch, while retaining the current 400 behavior for malformed refs.

**Comment by beardthelion** [comment]
Pushed `acb3214d` (two commits on current `main`).

**@ / revision shorthands (P1):** Confirmed `git check-ref-format --branch @` succeeds while `git merge @` is a no-op on the checked-out target. `validate_git_ref` now rejects `@`, any `@{...}` reflog form, and pseudoref names (`FETCH_HEAD`, `MERGE_HEAD`, `ORIG_HEAD`, `CHERRY_PICK_HEAD`, `RERERE_MERGE_HEAD`) before they reach storage or the diff/merge sinks. Boundary tests cover the full set; sink tests assert legacy rows with `@{-1}` / `FETCH_HEAD` / `MERGE_HEAD` return `git_error` (500) at merge with the PR still open, and `@{-1}` / `FETCH_HEAD` at diff.

**Spawn vs validation (P3):** `GitRefValidationError` splits `Invalid` (400 at `create_pr`/`create_repo`) from `GitUnavailable` (`AppError::Git`, 500). Unit test exercises the missing-binary path; HTTP tests on both storage boundaries assert `git_error` and no row written.

**Earlier P2 (HEAD, trailing dots, qualified refs):** Already on `799b73a9`; same tests remain green on this head.

**Rebase:** Merge-base is current `origin/main` (`bfc44f92`); no rebase required.

**Review by jatmn — CHANGES_REQUESTED** [review]
I found one architectural gap that explains the repeated review rounds. The option-injection fix (#378) and the HEAD/`@`/pseudoref false-merge fixes look solid on current head; what remains is finishing the branch-field contract in a way that does not require another denylist every time someone thinks of a new symbolic name.

## Merge readiness

- [P2] Confirm required test checks on the current PR head
  GitHub Actions run `33235270826` on `acb3214` reported `test (stable)` and `test (beta)` as failed while fmt/clippy, build, MSRV, and Docker smoke passed. You noted a green local run (829 tests). If those CI failures are infra flake, a green re-run is enough; if they are PR-owned, please fix before merge. I could not extract a specific failing test name from the published logs.

---

## Why this PR keeps getting review feedback

Each round has found another string class that passes the current validator but does not behave like a stored local branch name at the git sinks. That is not because the fixes so far are wrong — the layered defense (storage boundaries + sink guards, spawn-vs-400 split, mutation tests) is good work. The pattern is that the validator is still answering the wrong question.

**What the code validates today:** “Is this string valid *branch-name grammar*?” via `git check-ref-format --branch`, plus a growing explicit denylist (`refs/` prefix, `@`, `@{…}`, named pseudorefs).

**What the sinks actually need:** “When this string is passed as a bare git *revision argument*, will git treat it as the local branch `refs/heads/<stored-string>` and nothing else?”

Those are different questions. `check-ref-format --branch` only proves name syntax. Git’s revision parser still interprets many syntax-valid names as symbolic revisions at the sink:

| Stored value | Passes `check-ref-format --branch`? | What git does with the bare argv at diff/merge |
|---|---|---|
| `--output=/tmp/x` | No (now) | Was option injection (#378) — **fixed** |
| `HEAD`, `feature.` | No (now) | Symbolic / invalid branch — **fixed** |
| `@`, `@{-1}`, `FETCH_HEAD` | Mixed; now denied explicitly | False merge or symbolic ref — **fixed** |
| `refs/heads/main`, `refs/tags/v1` | Yes, but blocked by `refs/` rule | Qualified symbolic ref — **fixed** |
| `heads/main`, `tags/v1`, `origin/main` | Yes | Revision shorthands → `refs/heads/main`, `refs/tags/v1`, `refs/remotes/…` |
| `REBASE_HEAD`, `AUTO_MERGE`, … | Yes | Operation-state pseudorefs; merge/diff fail safely today |
| Literal branch `origin/feature` | Yes | Ambiguous if a remote-tracking ref with the same spelling exists |

Every review round has been another row in this table. Adding one more name to `PSEUDOREFS` or one more prefix rule closes that row but leaves the next row open. That is why feedback keeps dripping even as the security-critical issues get fixed.

The tests reinforce the same gap: boundary cases assert rejection of `refs/heads/main` and `refs/tags/v1` but not the equivalent bare shorthands `heads/main` and `tags/v1`, so the suite can go green while the symbolic-resolution class stays open.

---

## Root cause and recommended fix (close the class once)

**Root cause:** Branch fields are stored as short names but consumed as bare revision expressions (`git diff {target}...{source}`, `git worktree add … {target}`, `git merge {source}`). Validation guards grammar; git interprets semantics. Any syntax-valid string that is also a revision shorthand or pseudoref will keep surfacing until sinks stop relying on bare revision parsing.

**Recommended approach — make sinks use explicit local-branch refs:**

Keep storing short branch names in the DB (no API change). At the three sinks only, construct argv from `refs/heads/{name}` instead of the bare stored string:

- `branch_diff` / `branch_diff_names`: `refs/heads/{target}...refs/heads/{source}`
- `merge_branch` worktree: `git worktree add _merge_worktree refs/heads/{target}`
- `merge_branch` merge: `git merge refs/heads/{source}`
- post-merge `rev-parse`: already uses `refs/heads/{target}` today (`store.rs:991`) — this is why shorthand targets like `heads/main` can fail inconsistently after a detached worktree checkout

This is compatible with the PR’s note about not using `--` at the sink (that delimiter switches to pathspec mode; `refs/heads/…` is still a revision).

**Why this closes the class:**

1. **Option injection** — still blocked at storage (`check-ref-format` rejects leading `-`); `refs/heads/--output=…` is not a reachable path.
2. **HEAD / `@` / listed pseudorefs** — still blocked at storage by existing rules.
3. **`refs/…` qualified names** — still blocked at storage.
4. **`heads/main`, `tags/v1`, `origin/main` shorthands** — stored shorthand `heads/main` becomes `refs/heads/heads/main` at the sink (literal local branch name), not `refs/heads/main`. No silent retargeting to another ref namespace.
5. **Tag/remote ambiguity** — bare `v1` at the sink can resolve to a tag; `refs/heads/v1` resolves to the local branch head only.
6. **Denylist whack-a-mole** — you can keep `reject_revision_shorthand` for hygiene at the storage boundary, but sinks no longer depend on “this string must not be symbolic when bare.”

**Storage boundary can stay as-is** (or get simpler over time): `check-ref-format --branch`, leading-`-` rejection, explicit shorthand/pseudoref denylist, and `refs/` prefix rejection still fail fast with 400 and keep junk out of the DB. The sink guard remains defense-in-depth for legacy rows.

**Tradeoff to document:** A repo cannot use a PR branch field to point at a tag or a remote-tracking ref — only at a local branch. That matches the field names (`source_branch`, `target_branch`, `default_branch`) and the PR’s stated intent.

**Tests to add once (load-bearing, not another drip):**

- Boundary: `create_pr` / `create_repo` reject `heads/main` and `tags/v1` if you keep denying symbolic names at storage; otherwise rely on sink-prefix tests.
- Sink: with a stored literal branch `heads/main` (if allowed), prove diff/merge target `refs/heads/heads/main`, not `main`.
- Sink: repo with both `refs/heads/release` and `refs/tags/release` — stored `release` diff/merge uses the branch via `refs/heads/release`.
- Regression: existing poisoned-row / option-injection / false-merge tests stay green.

**What not to do:** Do not try to enumerate every git pseudoref and revision shorthand in a static list as the primary defense. `REBASE_HEAD`, `REVERT_HEAD`, `BISECT_HEAD`, and `AUTO_MERGE` are real holes in today’s list, but they fail safely (merge error, PR stays open) — fixing them one-by-one without sink-prefix semantics will invite the next name in the next review.

---

## Optional completeness (lower priority than sink-prefix)

If you prefer to keep bare revisions at the sinks for some reason, you would need a systematic “non-symbolic local branch only” policy at storage: full pseudoref set from git docs, revision-namespace prefixes (`heads/`, `tags/`, `remotes/`), and a documented decision on whether literal branch names like `origin/feature` are accepted knowing git may disambiguate them against remotes. That path tends to keep producing review feedback because git’s revision language is larger than branch-name grammar.

Separately, `fork_repo` copies `default_branch` without re-validation (`repos.rs:3106`). The PR correctly scoped that out (“fork takes no branch from the request”), and `create_pr` validates the resolved target, so this is not a #378 sink issue — only worth a line in docs or a follow-up if you want fork to inherit poisoned legacy defaults.

---

## Summary

The security work for #378 and the false-merge fixes look addressed. The remaining churn is architectural: **short names in the DB, bare revisions at the sink**. Switching sinks to `refs/heads/{name}` is the smallest change that matches the stored contract and should stop the drip. If you take that approach and add the consolidated tests above, I do not expect another round of “you missed this one symbolic string.”

**Review by beardthelion — COMMENTED** [review]
Withdrawn: self-review on own PR, posted in error.

**Comment by beardthelion** [comment]
@jatmn `d31042e` addresses the sink-prefix direction.

Sink argv now uses `refs/heads/{short_name}` in `branch_diff`, `branch_diff_names`, and `merge_branch` (worktree add and merge). Boundaries still store short names; `local_branch_ref` prefixes before every sink element.

Storage boundary also rejects `heads/`, `tags/`, and `remotes/` prefixes plus the extended pseudoref set. `create_pr` / `create_repo` call `validate_git_ref_with_git(&state.git_bin, …)`; spawn-unavailable tests set `state.git_bin` to a bad path (removed `GITLAWB_TEST_VALIDATE_GIT_BIN`).

Local runs on this head:
- `validate_git_ref` unit tests
- `create_pr_rejects_*` (option injection, `HEAD`, `@`, `heads/main`, `tags/v1`, `remotes/origin/main`)
- `create_repo_rejects_non_branch`
- spawn git-unavailable paths
- `poisoned_pr_row` / `legacy_pr_row` suites
- `branch_diff_names_lists`

Pre-push fmt and clippy passed on push.

**Review by jatmn — CHANGES_REQUESTED** [review]
I found issues that need to be addressed before this is ready.

## Findings

- [P1] Keep the merge worktree attached to the target branch
  `crates/gitlawb-node/src/git/store.rs:951`
  The new sink qualification is correct for revision arguments, but a worktree's final argument has different semantics: `git worktree add _merge_worktree refs/heads/{target_branch}` checks out the resolved commit in detached-HEAD mode instead of attaching the worktree to the local target branch. `git merge --no-ff refs/heads/{source_branch}` can therefore return success and create a merge commit only on that detached `HEAD`; after cleanup, `refs/heads/{target_branch}` still points to its pre-merge commit. The caller interprets that success as authoritative, persists the PR as `merged`, and emits `pull_request.merged`, leaving clients with a reported merge that is absent from the repository.

  Please address the root cause by separating the two contracts: retain explicit `refs/heads/{name}` qualification where Git consumes a revision, but create the temporary worktree in a way that attaches it to the selected local target branch. Then add an end-to-end successful-merge regression that starts with diverged target/source branches and asserts both that the target ref advances to a merge commit containing the source and that the PR status/webhook success path remains tied to that update. The existing negative/legacy-ref tests do not cover this normal successful merge path.

**Comment by beardthelion** [comment]
Confirmed the detached-worktree issue on `d31042e7`: `git worktree add` with `refs/heads/{target}` does not attach the worktree to the branch, so merge success did not move `refs/heads/{target}`.

`0c89c6b5` splits the contracts: worktree checks out the local branch name; `git merge` still takes `refs/heads/{source}`. Added `merge_pr_advances_target_ref_with_diverged_branches` (diverged main/feature, asserts ref advance + PR merged). Revert-check: putting `refs/heads/` back on worktree add REDs that test.

CodeQL on the prior head is unchanged; not part of this round.

**Review by jatmn — CHANGES_REQUESTED** [review]
I found one issue that needs to be addressed before this is ready.

## Findings

- [P1] Bind the temporary worktree to the exact local target branch
  `crates/gitlawb-node/src/git/store.rs:952`
  The source revision is now safely qualified as `refs/heads/{source_branch}`, but the target is passed to `git worktree add` as a bare name. `validate_git_ref` proves only that this string has valid branch-name grammar; it does not prove that `refs/heads/{target_branch}` exists or stop Git's revision DWIM rules from resolving the same name in another namespace. For example, with a `refs/tags/release` tag and no local `refs/heads/release` branch, a PR targeting `release` passes both boundary and sink validation. Git then creates `_merge_worktree` at the tag in detached-HEAD mode, the merge of the qualified source succeeds on that disposable HEAD, and cleanup discards the resulting commit without advancing any target branch. The final `rev-parse refs/heads/release` exits 128, but its status is not checked; because plain `rev-parse` echoes the unresolved token on stdout, `merge_branch` returns `"refs/heads/release"` as an apparently successful merge SHA. `merge_pr` consequently marks the row merged and emits `pull_request.merged` even though the requested target was never updated.

  Please close the resolution class rather than adding another symbolic-name denylist. Before creating the worktree, verify that the exact `refs/heads/{target_branch}` local ref exists, then create the worktree in attached-branch mode and verify its symbolic `HEAD` resolves to that same ref. After the merge, treat a nonzero `rev-parse --verify` (or any result that is not a commit for the exact target ref) as failure before updating the PR row or firing the webhook. Add an end-to-end regression with a same-named tag but no local target branch and assert that merge fails, the PR remains open, and no merged event is emitted; also retain a case where a same-named tag and local branch coexist and prove the local branch advances. This preserves the explicit `refs/heads/{source_branch}` source handling while making target selection and reported success refer to one unambiguous local branch.

**Review by jatmn — CHANGES_REQUESTED** [review]
I found issues that need to be addressed before this is ready.

## Merge readiness

- [P1] Resolve the failing CodeQL check before merge
  `CodeQL`
  The current PR check rollup reports CodeQL as failed while the other listed CI jobs pass. I could not retrieve the run log from the supplied run URL (GitHub returned 404), so its cause remains unknown; please rerun or resolve it and verify the result is green before merging this security-sensitive change.

## Findings

- [P2] Keep valid local branch names that start with a namespace word
  `crates/gitlawb-node/src/git/store.rs:828`
  `heads/release`, `tags/release`, and `remotes/release` are valid *local branch names*, not fully qualified refs: for example, `git check-ref-format --branch heads/release` succeeds and the branch is stored as `refs/heads/heads/release`. This new prefix check rejects those names before `check-ref-format` runs. Consequently, a repository that already has (or creates through another Git path) `refs/heads/heads/release` cannot create a PR using it; an existing PR with that name in either stored branch field also fails at the diff and merge sink guards.

  The root cause is conflating a short branch name such as `heads/release` with the revision shorthand `refs/heads/release`. The latter should remain disallowed as a fully qualified ref, but the former becomes unambiguous once this PR constructs `refs/heads/{name}`. Please remove the `heads/`, `tags/`, and `remotes/` prefix denylist, retain the `refs/` and symbolic/pseudoref rejections, and add regression coverage that creates a real `refs/heads/heads/release` branch and proves PR diff/merge target that branch rather than a tag or remote ref.

**Review by jatmn — APPROVED** [review]
lgtm

**Comment by beardthelion** [comment]
`1d08ae9c` fixes the leftover merge worktree. The CodeQL check is a separate matter and I do not think this branch can clear it; detail at the end.

**Leftover worktree.** `merge_branch` cleaned up `_merge_worktree` only when the directory still existed, but the state `git worktree add` refuses is the opposite one: a registration under `$GIT_DIR/worktrees` whose directory is gone. It fails with "missing but already registered", and every later merge on that repo fails with it. The state is self-inflicted, since the cleanup removed the directory even when the `git worktree remove` before it had failed.

Cleanup now runs unconditionally at entry, which also makes the earlier bail paths self-healing, and it passes `--force` twice. A single `--force` refuses a locked worktree and `git worktree prune` skips locked registrations, so with one flag a locked leftover survives the whole cleanup and `worktree add` then fails with "missing but locked worktree".

I enumerated the leftover states that can reach this code (registered or not, against directory present, gone, locked, emptied admin dir, and a regular file, under both default and `never` prune expiry). `remove -f -f` plus the directory removal clears all of them, and adding a `git worktree prune` changes no outcome in any of them, so there is no prune call in the final version.

Four tests. Three against `merge_branch`, for the stale registration, the locked variant, and an unregistered leftover directory (`git worktree remove` refuses that one with "is not a working tree", so the directory removal is what clears it). One through the merge endpoint on the locked state, asserting the PR row reaches merged, the reported `merge_sha` is the new `refs/heads/{target}`, the source commit is reachable, and nothing is left registered or on disk for the next merge. Each fixture asserts git really refuses the state before asserting recovery, so a future git that stopped refusing cannot quietly make them pass for the wrong reason.

**CodeQL.** Five of the eight alerts are in `crates/gitlawb-node/src/git/repo_store.rs` (lines 141, 336) and `crates/gitlawb-node/src/git/tigris.rs` (lines 239, 240, 242). Neither file is in this PR's diff, which is `api/mod.rs`, `api/pulls.rs`, `api/repos.rs`, `git/store.rs` and `test_support.rs`. The remaining three are in `store.rs` but inside `init_bare` at lines 9, 12 and 26, which this branch leaves byte-identical to `main`. All eight are open on `main` at `bfc44f92`.

The check also completes about nine minutes before the `Analyze (rust)` job it grades (on this head, 21:32:30 against 21:41:25), so its verdict on a given push is computed against the previous analysis. I pushed expecting a fresh run to clear it and it did not, which is what sent me to the file list. Clearing those alerts looks like separate work against `main` rather than anything this branch can do.

File: crates/gitlawb-node/src/git/store.rs [hunk 1]
```diff
@@ -799,18 +933,51 @@ pub fn merge_branch(
...
             .args(["worktree", "remove", "--force", "_merge_worktree"])
             .current_dir(repo_path)
             .output();
         let _ = std::fs::remove_dir_all(&worktree_path);
+    };
+
+    // Clean up any leftover worktree
+    if worktree_path.exists() {
+        remove_worktree();
+    }
```
[the remark this answers is not in the document]
Comment (beardthelion) [reply crates/gitlawb-node/src/git/store.rs]: Confirmed and fixed in `1d08ae9c`, though not with a prune.

Reproduced the state first: with the registration present and the directory gone, `git worktree add` fails exit 128 with "is a missing but already registered worktree", and the `if worktree_path.exists()` gate skipped the cleanup in exactly that case. The cleanup now runs unconditionally at entry.

On the prune specifically, I declined it after measuring. I enumerated the leftover states that can reach this code, registered or not, against directory present, gone, locked, emptied admin dir and a regular file, under both default and `never` prune expiry, and ran each against cleanup with and without a `git worktree prune`. The prune changed no outcome in any of them, because `git worktree remove --force --force` plus the directory removal already clears every one.

Chasing it did turn up a state the original cleanup missed, so the report was worth more than the specific remedy. A worktree that is both locked and missing survives a single `--force` ("cannot remove a locked working tree"), and `git worktree prune` skips locked registrations too, so `worktree add` then fails with "missing but locked worktree". The fix passes `--force` twice, which clears that and still handles the unlocked case.

Covered by three tests against `merge_branch` (stale registration, locked variant, unregistered leftover directory) plus one through the merge endpoint on the locked state.
